src/main/scala/org/neo4j/spark/dataframe/Neo4jDataFrame.scala

```diff
@@ -32,56 +35,66 @@ object Neo4jDataFrame {
         MERGE (target:${target._1.quote} {${targetLabel.quote} : row.target.${targetLabel.quote}}) ON CREATE SET target += row.target
         MERGE (source)-[rel:${relationship._1.quote}]->(target) ON CREATE SET rel += row.relationship
         """
-    val partitions = Math.max(1, (dataFrame.count() / 10000).asInstanceOf[Int])
     val config = Neo4jConfig(sc.getConf)
     dataFrame.repartition(partitions).foreachPartition(rows => {
-      val params: AnyRef = rows.map(r =>
-        Map(
-          "source" -> source._2.map(c => (renamedColumns.getOrElse(c, c), toJava(r.getAs(c)))).toMap.asJava,
-          "target" -> target._2.map(c => (renamedColumns.getOrElse(c, c), toJava(r.getAs(c)))).toMap.asJava,
-          "relationship" -> relationship._2.map(c => (c, toJava(r.getAs(c)))).toMap.asJava)
-          .asJava).asJava
-      execute(config, mergeStatement, Map("rows" -> params).asJava, write = true)
+      val driver: Driver = config.driver()
+      val session = driver.session(config.sessionConfig())
+      try {
+        rows.grouped(unwindBatchSize)
+          .foreach(chunk => {
+            val params: AnyRef = chunk.map(r => Map(
+              "source" -> source._2.map(c => (renamedColumns.getOrElse(c, c), toJava(r.getAs(c)))).toMap.asJava,
+              "target" -> target._2.map(c => (renamedColumns.getOrElse(c, c), toJava(r.getAs(c)))).toMap.asJava,
+              "relationship" -> relationship._2.map(c => (c, toJava(r.getAs(c)))).toMap.asJava)
+              .asJava).asJava
+            execute(session, mergeStatement, Map("rows" -> params).asJava, write = true)
+          })
+      } finally {
+        close(driver, session)
+      }
     })
   }
 
-
-  def createNodes(sc: SparkContext, dataFrame: DataFrame, nodes: (String, Seq[String]), renamedColumns: Map[String, String] = Map.empty): Unit = {
+  def createNodes(sc: SparkContext,
+                  dataFrame: DataFrame,
+                  nodes: (String, Seq[String]),
+                  renamedColumns: Map[String, String] = Map.empty,
+                  partitions: Int = 1,
+                  unwindBatchSize: Int = 10000): Unit = {
     dataFrame.cache()
     val nodeLabel: String = renamedColumns.getOrElse(nodes._2.head, nodes._2.head)
     val createStatement =
       s"""
         UNWIND $$rows as row
-        CREATE (node:${nodes._1.quote} {${nodeLabel.quote} : row.source.${nodeLabel.quote}})
-        SET node = row.node_properties
+        CREATE (node:${nodes._1.quote} {${nodeLabel.quote} : row.node_properties.${nodeLabel.quote}})
```

Comment:
possibly make this configurable via parameter (GH issue)

Reply:
MERGE vs. CREATE I mean

Reply:
I added it in this PR